docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+    response_model: BaseModel | BaseType, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call with response_model here
+```
+
+The first argument will be a Pydantic `BaseModel` or Python primative depending on what type `response_model` is.
+See `handle_call_response` section for information regarding the other arguments.
+
+For `BaseModel` you can grab the `CallResponse` via `response_model._response`.
+However for primatives `BaseType`, this information is not available.
+
+### `handle_structured_stream` and `handle_structured_stream_async`
+
+These functions will be called after streaming a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base._structured_stream import BaseStructuredStream
+
+def handle_structured_stream(
+    result: BaseStructuredStream,
+    fn: Callable,
+    context_manager: ContextManagerType | None,
+):
+```
+
+The first argument will be a Mirascope `StructuredStream` of the provider you are using.
+Like with `handle_stream`, `handle_structured_stream` will not be called until the `Generator` has been exhausted.
+
+### `custom_context_manager`
+
+You can define your own context manager where the yielded object is passed to each of the handlers.
+
+Here is an example using [SQLModel](https://sqlmodel.tiangolo.com/):
+
+```python
+from contextlib import contextmanager
+from typing import Any, Callable, Generator
+
+from sqlmodel import Session
+
+from mirascope.core.base import BaseCallResponse
+
+
+@contextmanager
+def custom_context_manager(
+    fn: Callable,
+) -> Generator[Session, Any, None]:
+    engine = ...
+    with Session(engine) as session:
+        yield session
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, session: Session | None
+):
+    # write to db 
+```
+
+In this example, the 3rd argument of your handler will be typed with the yielded `Session` and metadata can be written to your database of choice.
+
+### `custom_decorator`
+
+There may be existing libraries that have a decorator implemented already. You can pass that decorator in to `custom_decorator` which will wrap the Mirascope call with that decorator.
+
+## How to use your newly created decorator
+
+Now that you have defined your and created your `with_saving` decorator, you can wrap any Mirascope call, like so:
+
+```python
+from your_file import with_saving
+
+from mirascope.core import anthropic, prompt_template
+
+@with_saving
+@anthropic.call(model="claude-3-5-sonnet-20240620")
+@prompt_template("What is your purpose?")
+def run(): ...
+
+print(run())
+```
+
+In this example, when `run` is finished, `handle_call_response` will be called to collect the response.
+
+```python
+from typing import Callable
+from sqlmodel import Session
+
+from mirascope.core.base import BaseCallResponse
+
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, session: Session | None
```

Comment:
this confuses me, how can it be `session` shouldn't it be a context manager?

Reply:
The Session contextmanager yields a Session object.

Reply:
What confuses me is that the definition above has `context_manager: ContextManagerType | None` which makes it seem like `Session` is a `ContextManagerType`?

And then I don't see any `with ... as ...` which I generally expect from a context manager...

Reply:
its in `custom_context_manager`

Reply:
These functions are already wrapped in the context manager, so user doesnt do this for every handler.

Reply:
I'm honestly very confused here. Line 44 definition compared to this definition on line 179 just isn't making any sense to me. Perhaps it's just the way line 44 is defined / name? If line 179 is correct then 44 definitely needs fixing (as well as all of the other similar definitions across the various handlers). It's not clear that the context manager is run on their behalf.

Reply:
right its probably the name that is confusing you, will update it to make it clear